src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder.cs

```diff
@@ -527,29 +527,47 @@ bool IsDerivedType(TypeSymbol possibleDerived, TypeSymbol possibleBase)
             if (!recursive.Properties.IsDefault)
             {
                 // we have a "property" form
-                for (int i = 0; i < recursive.Properties.Length; i++)
+                foreach (var subpattern in recursive.Properties)
                 {
-                    var subPattern = recursive.Properties[i];
-                    Symbol? symbol = subPattern.Symbol;
-                    BoundPattern pattern = subPattern.Pattern;
-                    BoundDagEvaluation evaluation;
-                    switch (symbol)
+                    BoundPattern pattern = subpattern.Pattern;
+                    BoundDagTemp currentInput = input;
+                    if (subpattern.Symbols.IsEmpty)
                     {
-                        case PropertySymbol property:
-                            evaluation = new BoundDagPropertyEvaluation(pattern.Syntax, property, OriginalInput(input, property));
-                            break;
-                        case FieldSymbol field:
-                            evaluation = new BoundDagFieldEvaluation(pattern.Syntax, field, OriginalInput(input, field));
-                            break;
-                        default:
-                            RoslynDebug.Assert(recursive.HasAnyErrors);
-                            tests.Add(new Tests.One(new BoundDagTypeTest(recursive.Syntax, ErrorType(), input, hasErrors: true)));
-                            continue;
+                        RoslynDebug.Assert(recursive.HasAnyErrors);
```

Comment:
nit: `Debug.Assert` should work just as well now

Reply:
RoslynDebug.Assert can be removed wholesome any day now?

Reply:
Yes, it was useful when the BCL lacked some nullability attributes, but that has been resolved so we could remove now.